api/src/migrations/extract-enabled-locales.js

```diff
@@ -0,0 +1,47 @@
+const _ = require('lodash');
+
+const db = require('../db');
+const logger = require('../logger');
+
+module.exports = {
+  name: 'restrict-access-to-vault-db',
+  created: new Date(2023, 2, 23),
+  run: async () => {
+    const settings = await db.medic.get('settings');
+    if (settings.enabledLocales) {
+      return;
+    }
+
+    const translations = await db.medic.query('medic-client/doc_by_type', {
+      startkey: [ 'translations', false ],
+      endkey: [ 'translations', true ],
+      include_docs: true
+    });
+    const translations2 = await db.medic
+      .allDocs({ startkey: 'messages-', endkey: 'messages-\ufff0', include_docs: true });
+    console.log("translations", translations);
+    console.log("translations2", translations2);
+    const enabledLocales = [];
+    for (const row of translations.rows) {
+      if (validTranslationsDoc(row.doc)) {
+        enabledLocales.push(row.doc.code);
+      }
+    }
+    settings.enabledLocales = enabledLocales;
+    // await db.medic.put(settings);
+  }
+};
+
+const validTranslationsDoc = doc => {
+  if (!doc || doc.type !== 'translations' || !doc.code) {
+    return false;
+  }
+
+  if (_.isObject(doc.generic) || _.isObject(doc.values)) {
+    return true;
+  }
+
+  logger.warn(`Failed to load translations for "${doc.code}"("${doc.name}"). Translations document malformed.`);
+  return false;
+};
+
```

Comment:
As [mentioned in this comment](https://github.com/medic/cht-core/issues/6281#issuecomment-1480622151) I'm not convinced that a migration will work here.

An alternative approach would be to change all the code that...
1. Looks in the settings for `enabledLocales` and if it's found, use it
2. Falls back to the current code when it's not found.

Another option would be to find all languages that are enabled in settings _or_ in the translation doc but that makes it harder to deprecate and remove the old way.

Reply:
> a first migration that sets the enabledLocales in settings that we release now~ish -- with the fallback logic

I don't think the migration adds anything. What I expect will happen is...

1. During the upgrade the migration will run and update the `settings` doc.
2. Now anyone looking for enabled locales will not need the fallback logic - it'll just work.
3. Soon after the app dev uploads the settings from CI or CLI and overwrites the `settings` doc which removes the `enabledLocales` property.
4. Now we're back to where we started and the fallback logic will run every time

I think it's less risky to just write the fallback logic and not bother with the migration.

Reply:
Fair point, that would render the migration useless...

I'll move forward with just the fallback logic and open another issue to deprecate the old way for 5.x

Reply:
Opened up the follow up issue medic/cht-core#8157